sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/CustomTimestampPolicyWithLimitedDelay.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.rabbitmq;
+
+import java.util.Optional;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+
+/**
+ * A policy for custom record timestamps where timestamps within the queue are expected to be
+ * roughly monotonically increasing with a cap on out of order event delays (say 1 minute). The
+ * watermark at any time is '({@code Min(now(), Max(event timestamp so far)) - max delay})'.
+ * However, watermark is never set in future and capped to 'now - max delay'. In addition, watermark
+ * advanced to 'now - max delay' when the queue has caught up (previous read attempt returned no
+ * message)
+ *
+ * @see "org.apache.beam.sdk.io.kafka.CustomTimestampPolicyWithLimitedDelay for the inspiration and
+ *     corresponding KafkaIO policy"
+ */
+public class CustomTimestampPolicyWithLimitedDelay extends TimestampPolicy {
+  private final Duration maxDelay;
+  private final SerializableFunction<RabbitMqMessage, Instant> timestampFunction;
+  private Instant maxEventTimestamp;
+
+  /**
+   * A policy for custom record timestamps where timestamps are expected to be roughly monotonically
+   * increasing with out of order event delays less than {@code maxDelay}. The watermark at any time
+   * is {@code Min(now(), max_event_timestamp) - maxDelay}.
+   *
+   * @param timestampFunction A function to extract timestamp from the record
+   * @param maxDelay For any record in the Kafka partition, the timestamp of any subsequent record
```

Comment:
I see this was copied from `Kafka` :) but we should probably change this to "Rabbit queue"
```suggestion
   * @param maxDelay For any record in the Rabbit queue, the timestamp of any subsequent record
```

Reply:
ha! thanks